lib/bakeAmbientOcclusion.js

```diff
@@ -11,52 +11,428 @@ var Matrix3 = Cesium.Matrix3;
 var Matrix4 = Cesium.Matrix4;
 var Quaternion = Cesium.Quaternion;
 var Ray = Cesium.Ray;
+var ShaderSource = Cesium.ShaderSource;
 var WebGLConstants = Cesium.WebGLConstants;
+
+var byteLengthForComponentType = require('./byteLengthForComponentType');
 var clone = require('clone');
 var Jimp = require('jimp');
 var NodeHelpers = require('./NodeHelpers');
 var readAccessor = require('./readAccessor');
+var GeometryMath = require('./GeometryMath');
 
 module.exports = {
     bakeAmbientOcclusion: bakeAmbientOcclusion,
+    generateOptions: generateOptions,
     generateRaytracerScene: generateRaytracerScene,
-    computeAmbientOcclusionAt: computeAmbientOcclusionAt
+    computeAmbientOcclusionAt: computeAmbientOcclusionAt,
+    raytraceAtTriangleCenters: raytraceAtTriangleCenters,
+    raytraceOverTriangleSamples: raytraceOverTriangleSamples
 };
 
 var scratchRay = new Ray();
-var cartesian3Scratch1 = new Cartesian3();
-var cartesian3Scratch2 = new Cartesian3();
-var cartesian3Scratch3 = new Cartesian3();
+var barycentricCoordinateScratch = new Cartesian3();
+var worldPositionScratch = new Cartesian3();
+var worldNormalScratch = new Cartesian3();
 var cartesian2Scratch = new Cartesian2();
 var quaternionScratch = new Quaternion();
 var matrix3Scratch = new Matrix3();
 var matrix4Scratch = new Matrix4();
 
-function bakeAmbientOcclusion(gltf, options) {
+// AO options
+// Basic:
+// - toTexture: Bake AO to existing diffuse textures instead of to vertices. Does not modify shaders. Default: false
+// - groundPlane: Add a groundplane at the lowest point of the model for AO computation. Default: false
+// - ambientShadowContribution: amount of AO to show when blending between shader computed lighting and AO. 1.0 is full AO, 0.5 is a 50/50 blend. Default: 0.5
+// - quality: Valid settings are high, medium, and low. Default: low
+
+// Advanced quality settings:
+// - density: for vertex baking mode, sample count per unit length at which to generate additional sample points. Default: 4.0, 2.0, 1.0
+// - resolution: for texture baking mode, resolution at which to generate an AO texture. AO texture will be scaled to diffuse texture. Default: 512, 256, 128.
+// - numberRays: number of rays to cast from each sample point. Clamps to nearest smaller square. Default: 64, 36, 16
+// - triangleCenterOnly: for vertex baking mode, do not sample a triangle's area in addition to its center using a grid. Default: false
+
+// Advanced appearance settings:
+// - rayDistance: ray bottom-out distance in world-space units. Default: 1.0
+// - nearCull: near cull depth for ray-triangle collisions. Default: 0.0001 units in world space.
+// - shaderMode: how should the shader use per-vertex AO? Valid settings are multiply, replace, and blend. Default: blend
+// - scene: which scene to use when baking AO. Default: gltf default
+
+function bakeAmbientOcclusion(gltf, aoOptions) {
```

Comment:
Throughout, why pass separate `gltf` and `aoOptions`.  Why not just pass `options` like Cesium does?  https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide#options-parameters